### DISCUSSION TITLE
Read configuration from device

### DISCUSSION THREAD

**Discussion by kamaaina** [kriomant/ch57x-keyboard-tool#64] [Ideas]
Currently, the upload (aka programming) of device is supported. Would it be possible to read the configuration on the device and report that back? 

Perhaps something along the lines of:
```
./ch57x-keyboard-tool configuration > current_mapping.yml
``` 

This would be useful to see how the device is currently programmed.

I'd be willing to help with this feature request as I have several devices, but I would need some guidance on where to start. Obviously, wireshark and the USB capture would yield this information, but there is a command that can be sent for the device to return its settings? If not, any suggestions on how this could be obtained?

**Comment by kriomant** [comment]
I'm not sure it is possible at all. I'm not aware any formal specification or data sheet for such keyboards.
So the only way to find out how to do it is to find existing software which can do it.

**Comment by kamaaina** [comment]
i think i might have found a way to do this... i do not understand every single byte, but i don't think that is actually necessary to get the desired result. 

i forked @pferreir branch where he added support for 1189:8840 vid/pid and will be adding code there. once i get it all working (i have a few devices to test with), would you be open to merging a PR for this?

**Reply by kriomant** [reply]
Of course

**Reply by juliocargnin** [reply]
> i think i might have found a way to do this... i do not understand every single byte, but i don't think that is actually necessary to get the desired result.
> 
> i forked @pferreir branch where he added support for 1189:8840 vid/pid and will be adding code there. once i get it all working (i have a few devices to test with), would you be open to merging a PR for this?

Did you find a way to extract the .yaml from the device? I have lost my configuration in a harddrive crash so I'd like to recover my configuration from the device itself so I can continue customizing my keys from where it is.

**Comment by kamaaina** [comment]
i am not sure if @kriomant integrated changes yet, but you are welcome to use my tool here - https://github.com/kamaaina/macropad_tool 

however, i ended up changing the config from yml to ron format. you can connect your device and run:
`macropad-tool read`

if your device supports reading, the configuration will be written to stdout

**Reply by juliocargnin** [reply]
I was able to retrieve the stored information from the device with your tool! Thank you!!
